File: Recipe - Java/app/src/main/java/com/example/recipe_java/ui/launcher/RestaurantAdapter.java
package com.example.recipe_java.ui.launcher;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.example.recipe_java.Data;
import com.example.recipe_java.R;
import com.example.recipe_java.models.Restaurant;
import org.jetbrains.annotations.NotNull;

public class RestaurantAdapter extends RecyclerView.Adapter<RestaurantAdapter.RestaurantViewHolder> {

    private RestaurantClicked listener;

    public RestaurantAdapter(RestaurantClicked listener) {
        this.listener = listener;
    }

    @NonNull
    @NotNull
    @Override
    public RestaurantViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_restaurant, parent, false);
        return new RestaurantAdapter.RestaurantViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull RestaurantViewHolder holder, int position) {
        Restaurant restaurant = Data.restaurants.get(position);
        holder.restaurantName.setText(restaurant.getName());
        holder.restaurantAddress.setText(restaurant.getAddress());

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            // Example: sealed event
            public void onClick(View view) {
                listener.onRestaurantClicked(restaurant);
            }
        });
    }

    @Override
    public int getItemCount() {
        return Data.restaurants.size();
    }

    class RestaurantViewHolder extends RecyclerView.ViewHolder {

        TextView restaurantName;
        TextView restaurantAddress;

        public RestaurantViewHolder(@NonNull @NotNull View itemView) {
            super(itemView);
            restaurantName = itemView.findViewById(R.id.restaurantNameText);
            restaurantAddress = itemView.findViewById(R.id.restaurantAddressText);
        }
    }
}

File: Recipe - Kotlin/app/src/main/java/com/example/recipe/ui/launcher/RestaurantAdapter.kt
package com.example.recipe.ui.launcher

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.recipe.Data
import com.example.recipe.R
import com.example.recipe.models.Restaurant
import com.example.recipe.sealedEvents.RestaurantClicked

class RestaurantAdapter(
    private val event: (RestaurantClicked) -> Unit
): RecyclerView.Adapter<RestaurantAdapter.RestaurantViewHolder>() {

    private var restaurants: MutableList<Restaurant> = mutableListOf()

    // Example: predefined argument
    fun setRestaurants(restaurants: MutableList<Restaurant> = Data.restaurants) {
        this.restaurants = restaurants
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RestaurantViewHolder {
        val view: View = LayoutInflater.from(parent.context).inflate(R.layout.item_restaurant, parent, false)
        return RestaurantViewHolder(view)
    }

    override fun onBindViewHolder(holder: RestaurantViewHolder, position: Int) {
        val restaurant = restaurants[position]
        holder.restaurantName.text = restaurant.name
        holder.restaurantAddress.text = restaurant.address

        holder.itemView.setOnClickListener {
            // Example: sealed event
            event.invoke(RestaurantClicked.RestaurantPicked(restaurant))
        }
    }

    override fun getItemCount(): Int {
        return restaurants.size
    }

    inner class RestaurantViewHolder(itemView: View): RecyclerView.ViewHolder(itemView) {
        var restaurantName: TextView = itemView.findViewById(R.id.restaurantNameText)
        var restaurantAddress: TextView = itemView.findViewById(R.id.restaurantAddressText)
    }
}
